lib/vector-core/src/event/ser.rs

```diff
@@ -0,0 +1,119 @@
+use bytes::{Buf, BufMut};
+use enumflags2::{bitflags, BitFlags, FromBitsError};
+use prost::Message;
+use snafu::Snafu;
+use vector_buffers::encoding::{AsMetadata, Encodable};
+
+use super::{proto, Event, EventArray};
+
+#[derive(Debug, Snafu)]
+pub enum EncodeError {
+    #[snafu(display("the provided buffer was too small to fully encode this item"))]
+    BufferTooSmall,
+}
+
+#[derive(Debug, Snafu)]
+pub enum DecodeError {
+    #[snafu(display(
+        "the provided buffer could not be decoded as a valid Protocol Buffers payload"
+    ))]
+    InvalidProtobufPayload,
+    #[snafu(display("unsupported encoding metadata for this context"))]
+    UnsupportedEncodingMetadata,
+}
+/// Flags for describing the encoding scheme used by our primary event types that flow through buffers.
+///
+/// # Stability
+///
+/// This enumeration should never have any flags removed, only added.  This ensures that previously
+/// used flags cannot have their meaning changed/repurposed after-the-fact.
+#[bitflags]
+#[repr(u32)]
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum EventEncodableMetadataFlags {
+    /// Chained encoding scheme that first tries to decode as `EventArray` and then as `Event`, as a
+    /// way to support gracefully migrating existing v1-based disk buffers to the new
+    /// `EventArray`-based architecture.
+    ///
+    /// All encoding uses the `EventArray` variant, however.
+    DiskBufferV1CompatibilityMode = 0b1,
```

Comment:
@tobz Should there be a `EventArray`-only mode, or am I misunderstanding the purpose of these flags?

Reply:
In the future: yes, we'll transition to adding a new mode that is `EventArray` only.

Right now, this encoder reports that it specifically uses this mode for encoding, and that it supports decoding it.  The intention is that the most up-to-date mode is the one an encoder will report as using for decoding, but that it may report that it can handle decoding more than one mode.

Since we already encode as `EventArray` with this one, adding another specific one doesn't change how we encode, but does make it easier to make a mistake where the decoding stuff in disk buffer v1 says that the payload is in "EventArrayOnly" mode, or something, and then decoding breaks.

This way, we're basically codifying the behavior for all disk buffers -- encode as `EventArray`, but decode as `EventArray` or `Event` -- by having a flag that specifies that, and _only having_ that flag.